docs: explain the situation with --manual renewal

alexzorin opened the pull request:
Users keep hitting this. This is the documentation half of #6280. On top of this, we still need to add some CLI output that achieves the same thing.

I can't remember why I closed #8601. Our meeting notes on that day just say:

> - Manual that doesn’t renew -- let’s try to integrate throughout, where it’s relevant
>    - Update documentation
>    - Add a message in the renewal information message
>    - Possibly contact prominent tutorials, such as the amazon one
>       - https://lightsail.aws.amazon.com/ls/docs/en_us/articles/amazon-lightsail-using-lets-encrypt-certificates-with-wordpress We can submit a PR!
>    - Potential future: update --manual to be --manual-without-renewal or something, make --manual hidden

Screenshot:

![image](https://user-images.githubusercontent.com/311534/122488524-f998be80-d020-11eb-91d7-d3d175ef1ad1.png)

![image](https://user-images.githubusercontent.com/311534/122488556-0ddcbb80-d021-11eb-9745-bba3099d23ac.png)

ohemorange commented:
> I can't remember why I closed #8601.

Based on our meeting notes, I suspect I rebelled against the idea of adding an extra interactive prompt that users need to click through to get to what they want to do. So rather than stopping them in the middle of when they're getting a certificate, put it in the instructions they're reading (whether that's these docs or a tutorial), and at the point when we otherwise give users information about how to set up renewal (after they've gotten the cert).

Which makes me think, maybe we want to link to this anchor point from our other renewal information.

ohemorange commented on certbot/docs/using.rst:
```diff
@@ -229,11 +229,14 @@ For example, for the domain ``example.com``, a zone file entry would look like:
+to automatically set up the required HTTP and/or TXT challenges.
+
+To renew a certificate using ``--manual`` without hooks, repeat the same ``certbot --manual``
+command you used to create the certificate originally.
```
Maybe put something about "you can't just stick that in a cron job because you're going to have to manually update the dns probably", to cover the confusion that came up in #6280?

alexzorin replied:
Good idea.

![image](https://user-images.githubusercontent.com/311534/122139337-18198100-ce8c-11eb-9bc3-244c18245d0f.png)

alexzorin commented:
> Which makes me think, maybe we want to link to this anchor point from our other renewal information.

Do you mean in the user guide, or from the instructions generator?

I think the "Renewing certificates" section in the user guide would benefit from a bit of a clean up/rewrite anyway. It's very long and a bit kitchen-sinky.

ohemorange commented:
> Do you mean in the user guide, or from the instructions generator?

I meant in the user guide. I was thinking in the section we just added that's linked from the terminal output, but actually now that I think about it more, we're not linking there from here so maybe it's not the right place.

Yes I agree that section could use a clean-up.

But either way what I'm trying to get at is that however they ended up using `--manual`, they should be directed to the right information. If we think that the only place they follow these instructions from is the section being edited here that's fine and we can leave it, that's fine, I just want to make sure we think through the flows of all the options.

alexzorin commented:
> But either way what I'm trying to get at is that however they ended up using --manual, they should be directed to the right information

OK great, this is where my head is too. I just wasn't sure what flow would lead them to that section.

With the instructions generator, I was thinking about users who are arriving to `#manual` from Step 1 of https://certbot.eff.org/lets-encrypt/debianbuster-nginx#wildcard. They should already see the renewal caveat if they do that.

A minority of users might end up at `#automated-renewal` if they don't have `--preconfigured-renewal` set. This is an even bigger problem for the majority of the users who do have it set, because we just tell them that the certificate will be autorenewed full stop, which is untrue. I very much think that situation requires a more drastic CLI change, which links directly to the manual renewal caveat in the user guide.

On top of that, we can *also* call out the manual plugin at the beginning of `#renewal`. I think the cost of doing so now is relatively low so I just did it. We can revisit it with the `--manual` CLI change:

![image](https://user-images.githubusercontent.com/311534/122301806-8883d900-cf44-11eb-9c7b-68d8fe60082b.png)

ohemorange commented:
> A minority of users might end up at #automated-renewal if they don't have --preconfigured-renewal set. This is an even bigger problem for the majority of the users who do have it set, because we just tell them that the certificate will be autorenewed full stop, which is untrue. I very much think that situation requires a more drastic CLI change, which links directly to the manual renewal caveat in the user guide.

Wait hold on, are users with it set getting the message even if they use manual? I agree, we should definitely fix that (when we make non-docs changes), and I like that solution.

Then do we also need to tell people somewhere (probably here) where to find the renewal command we set up so they can add the auth hook flag, or how to create one if they don't already have one? Because we half mention it here, then direct them to hooks, which doesn't say how to set up/modify renewal. I feel like we're creating a choose your own adventure novel here, which is not how docs pages should be, but then maybe in that upcoming rewrite we just need to explicitly and independently define the different aspects so we can always just pop in the correct explanation as needed. There may not be a clean way to do this without that rewrite, if we don't want to just make all our explanations more of a mess than they already are. But leaving users without anything feels bad.

alexzorin commented:
> Wait hold on, are users with it set getting the message even if they use manual? I agree, we should definitely fix that (when we make non-docs changes), and I like that solution.

The part of Certbot that reports on next steps doesn't KNOW that this particular combination of `authenticator=manual,manual-auth-hook=None` doesn't support autorenewal. I think we could probably just hard-code in that case so whatever.

>Then do we also need to tell people somewhere (probably here) where to find the renewal command we set up so they can add the auth hook flag, or how to create one if they don't already have one?

I think in the latest changes to the renewal cron instructions, we came to the conclusion that we can't just tack on flags to `certbot renew`, because sometimes it's either not possible or squirreled away in some esoteric place (snapd renewal systemd unit, Windows Scheduled Task).

If somebody wants to migrate from manual-without-hooks to manual-with-hooks, changing renewal params is what they need to do, yeah. But, in a UX sense, this option doesn't really exist. It's either force a live renewal or modify the renewal params file and do a `--dry-run`. I don't super want to get into that can of worms right now because it's hard to solve.

The intention of this PR is to give a heads-up to users who were planning to use the manual plugin under the impression that autorenewal will be possible.

I think the situation itself is messy and should be fixed, but imoooo this is a marginal improvement over saying nothing and then Certbot throwing a surprise party in 90 days.

ohemorange commented:
> I think in the latest changes to the renewal cron instructions, we came to the conclusion that we can't just tack on flags to certbot renew, because sometimes it's either not possible or squirreled away in some esoteric place (snapd renewal systemd unit, Windows Scheduled Task). Or do you mean into cli.ini and have it affect all certificates?

No, that is in fact exactly what I meant. But I actually assumed that we wouldn't store the auth hook in the renewal config and so it wouldn't get called, and was trying to mitigate that. But I do agree with you, we should really just warn people that did it without the hook that it won't renew. Is there a clear, concise way that we can say "If you want the auto-renew functionality, create a new certificate using a manual auth hook"? Is that too complicated to append to what you have here as an alternate option? Hm, actually, if someone runs it again and we suggest that they renew a cert rather than creating a new one, is it going to the the thing where it fails to save the new cli options to the renewal file like you recently found in https://github.com/certbot/certbot/issues/8906? I bet it is.

> The intention of this PR is to give a heads-up to users who were planning to use the manual plugin under the impression that autorenewal will be possible.
>
> I think the situation itself is messy and should be fixed, but imoooo this is still a real improvement over saying nothing and then Certbot throwing a surprise party in 90 days.

Yes, that makes sense, I'm just wary of feeling like we've solved something by throwing it into the docs, since we've definitely hit that before where it's still confusing in practice. And specifically since we talked a bunch recently about making sure to have actionable next steps, I'd like to do that to the extent reasonable here.

alexzorin commented:
>Is there a clear, concise way that we can say "If you want the auto-renew functionality, create a new certificate using a manual auth hook"?

I originally had more text like this which was intended to divert the user from the manual plugin altogether:

*If you can use one of the other plugins which support autorenewal to create your certificate, doing so is highly recommended.*

Realistically, `--manual-auth-hook` is functionality for experts. It takes sooo much knowledge, especially to do it right (consider the complexity of multiple challenges on one DNS label). IMO it is not a first resort for anyone.

If I understand your feedback right, I can propose these changes to the "Renewal with the manual plugin" section:

- [ ] Add the statement diverting users to other plugins
- [ ] Show how to use `--force-renewal` to upgrade from manual-without-hook to manual-with-hook (also, see my discussion below)

> Hm, actually, if someone runs it again and we suggest that they renew a cert rather than creating a new one, is it going to the the thing where it fails to save the new cli options to the renewal file like you recently found in #8906? I bet it is.

Well, a renewal or force-renewal with an added `--manual-auth-hook` will do the right thing and update the renewal params. 

But of course the problem is updating those params in a sane way (editing `.conf` files is dangerous, and `--force-renewal` is a waste of rate limits).

---

This is super off-topic but besides the renewal actions of:

1. keep the cert (noop=certonly/reinstall=run)
2. renew the cert (`--force-renewal`)

it would be really nice to have something like:

3. dry-run renew the cert and update the renewal params if the dry-run succeeds (`--dry-run-and-update`?)

then it would be easier to just say in the documentation:

>You can upgrade your --manual cert to support autorenewal by adding an authentication hook:
>
>     certbot renew --cert-name manual.example.com --manual-auth-hook /path/to/hook.sh --dry-run-and-update

ohemorange commented:
> If I understand your feedback right, I can propose these changes to the "Renewal with the manual plugin" section:
>
> Add the statement diverting users to other plugins
> Show how to use --force-renewal to upgrade from manual-without-hook to manual-with-hook (also, see my discussion below)

You did understand right, but actually you've convinced me, and now I think the first will suffice.

> It would be really nice to have something like:
>
> dry-run renew the cert and update the renewal params if the dry-run succeeds (--dry-run-and-update?)

It sure would! I think there might be an open issue about this somewhere, actually?

alexzorin commented:
Okay, added that one statement and updated screenshots. I also took the chance to rewrite the description in the plugins table, to highlight that the manual plugin has restrictions surrounding autorenewal.

>It sure would! I think there might be an open issue about this somewhere, actually?

There's a couple: https://github.com/certbot/certbot/issues/8822. I think I gave it a shot once, but ran into some roadblock with reconstitution.

ohemorange commented on certbot/docs/using.rst:
```diff
@@ -57,10 +57,11 @@ standalone_ Y    N    | Uses a "standalone" webserver to obtain a certificate.
+manual_     Y    N    | Obtain a certificate by manually following instructions to    http-01_ (80) or
+                      | perform domain validation yourself. Certificates created this dns-01_ (53)
+                      | way do not support autorenewal.
+                      | Autorenewal may be enabled by specifying scripts to automate
```
I generally like this, but how do you feel about either using the same name as we use to refer to it elsewhere, or just literally saying the flag name? To make it easier to cross-reference.

alexzorin replied:
Eeeeeeeee you're right. This isn't the only place I'm being inconsistent.

So, what's the real name? 

In the documentation, the header is *Pre and Post Validation Hooks*.

Personally I always call them authentication hooks, but I am willing to fix that habit, if we're calling them validation hooks.

ohemorange replied:
I think we call them authentication hooks more often, given a quick control-f around the docs, and looking through the implementation. But either way I do like using the name that matches the flag name, if not the flag name itself.

ohemorange reviewed (changes requested):
Loving the additions, one suggestion on the new table text.
